### PR TITLE
Introduce support for `async gen` blocks

### PR DESCRIPTION
**Pull request by compiler-errors**
I'm delighted to demonstrate that `async gen` block are not very difficult to support. They're simply coroutines that yield `Poll<Option<T>>` and return `()`.

~~**This PR is WIP and in draft mode for now** -- I'm mostly putting it up to show folks that it's possible. This PR needs a lang-team experiment associated with it or possible an RFC, since I don't think it falls under the jurisdiction of the `gen` RFC that was recently authored by oli (https://github.com/rust-lang/rfcs/pull/3513, https://github.com/rust-lang/rust/issues/117078).~~

### Technical note on the pre-generator-transform yield type:

The reason that the underlying coroutines yield `Poll<Option<T>>` and not `Poll<T>` (which would make more sense, IMO, for the pre-transformed coroutine), is because the `TransformVisitor` that is used to turn coroutines into built-in state machine functions would have to destructure and reconstruct the latter into the former, which requires at least inserting a new basic block (for a `switchInt` terminator, to match on the `Poll` discriminant).

This does mean that the desugaring (at the `rustc_ast_lowering` level) of `async gen` blocks is a bit more involved. However, since we already need to intercept both `.await` and `yield` operators, I don't consider it much of a technical burden.

r? @ghost

### CODE REVIEW

**Comment by compiler-errors** [comment]
Importantly, this still needs a test 💀 

I'll come up with one that uses a no-op waker or something.

**Comment by traviscross** [comment]
@rustbot labels +I-lang-nominated

Nominating based on discussion to confirm that we're happy for `async gen` to land in nightly under the same experimental auspices as for `gen`.

**Comment by compiler-errors** [comment]
I believe this is ready for review

r? @eholk, though please reassign if you're busy

File: compiler/rustc_mir_transform/src/coroutine.rs [hunk 1]
```diff
@@ -1331,7 +1393,8 @@ fn create_coroutine_resume_function<'tcx>(
 
     if can_return {
         let block = match coroutine_kind {
-            CoroutineKind::Async(_) | CoroutineKind::Coroutine => {
+            // FIXME(gen_blocks): Should `async gen` yield `None` when resumed once again?
```
Comment (eholk) [inline compiler/rustc_mir_transform/src/coroutine.rs:1396]: I tend to think no, but I can see arguments either way. Something to figure out for the RFC, I guess.

File: compiler/rustc_parse/src/parser/expr.rs [hunk 2]
```diff
@@ -2261,9 +2262,16 @@ impl<'a> Parser<'a> {
...
-        if let Some(CoroutineKind::Async { span, .. }) = asyncness {
-            // Feature-gate `async ||` closures.
-            self.sess.gated_spans.gate(sym::async_closure, span);
+        match coroutine_kind {
+            Some(CoroutineKind::Async { span, .. }) => {
+                // Feature-gate `async ||` closures.
+                self.sess.gated_spans.gate(sym::async_closure, span);
+            }
+            Some(CoroutineKind::AsyncGen { span, .. }) | Some(CoroutineKind::Gen { span, .. }) => {
+                // Feature-gate `async ||` closures.
```
Comment (eholk) [inline compiler/rustc_parse/src/parser/expr.rs:2271]:
```suggestion
                // Feature-gate `async gen ||` closures.
```

**Review by eholk — APPROVED** [review]
Looks good. Just some minor questions/suggestions that you can take or not. Nothing blocking, so feel free to r=me after making any changes you want to take.

File: compiler/rustc_ast_lowering/src/lib.rs [hunk 3]
```diff
@@ -1960,20 +1968,21 @@ impl<'a, 'hir> LoweringContext<'a, 'hir> {
...
-            bindings: arena_vec![self; self.assoc_ty_binding(assoc_ty_name, span, output_ty)],
+            bindings: arena_vec![self; self.assoc_ty_binding(assoc_ty_name, opaque_ty_span, output_ty)],
             parenthesized: hir::GenericArgsParentheses::No,
             span_ext: DUMMY_SP,
         });
 
         hir::GenericBound::LangItemTrait(
             trait_lang_item,
-            self.lower_span(span),
+            opaque_ty_span,
```
[the remark this answers is not in the document]
Comment (compiler-errors) [reply compiler/rustc_ast_lowering/src/lib.rs:1985]: No, it has been lowered

File: library/core/src/async_iter/async_iter.rs [hunk 4]
```diff
@@ -109,3 +110,28 @@ where
...
+        Poll::Ready(Some(t))
+    }
+
+    /// A helper function for internal desugaring -- produces `Pending`,
+    /// which corresponds to the async iterator pending on an `.await`.
+    #[unstable(feature = "async_gen_internals", issue = "none")]
+    #[cfg_attr(not(bootstrap), lang = "AsyncGenPending")]
+    pub fn async_gen_pending() -> Self {
+        Poll::Pending
+    }
```
[the remark this answers is not in the document]
Comment (compiler-errors) [reply library/core/src/async_iter/async_iter.rs]: Yes, but in the process, I found a bug in hir typeck.

**Comment by compiler-errors** [comment]
Rebased

@bors r=eholk

**Comment by compiler-errors** [comment]
I have literally no idea why this would be a perf regression. Any ideas, @rust-lang/wg-compiler-performance?

**Comment by Kobzol** [comment]
```
<  15,266,505  ???:
  -19,862,558    <hashbrown::raw::RawTable<(rustc_span::def_id::LocalDefId, rustc_hir::hir_id::ItemLocalId)>>::reserve_rehash::<hashbrown::map::make_hasher<rustc_span::def_id::LocalDefId, rustc_hir::hir_id::ItemLocalId, core::hash::BuildHasherDefault<rustc_hash::FxHasher>>::{closure#0}>
   19,850,279    <hashbrown::raw::RawTable<(rustc_ast::node_id::NodeId, rustc_hir::hir_id::ItemLocalId)>>::reserve_rehash::<hashbrown::map::make_hasher<rustc_ast::node_id::NodeId, rustc_hir::hir_id::ItemLocalId, core::hash::BuildHasherDefault<rustc_hash::FxHasher>>::{closure#0}>
    5,150,361    rustc_query_system::dep_graph::graph::hash_result::<&rustc_middle::ty::typeck_results::TypeckResults>
    2,563,601    <rustc_middle::ty::Ty as rustc_serialize::serialize::Decodable<rustc_middle::query::on_disk_cache::CacheDecoder>>::decode
    1,671,252    <&rustc_middle::ty::list::List<rustc_middle::ty::generic_args::GenericArg> as rustc_serialize::serialize::Decodable<rustc_middle::query::on_disk_cache::CacheDecoder>>::decode
    1,670,754    rustc_query_impl::plumbing::encode_query_results::<rustc_query_impl::query_impl::typeck::QueryType>::{closure#0}
    1,651,399    rustc_middle::ty::codec::encode_with_shorthand::<rustc_middle::query::on_disk_cache::CacheEncoder, rustc_middle::ty::Ty, <rustc_middle::query::on_disk_cache::CacheEncoder as rustc_type_ir::codec::TyEncoder>::type_shorthands>
    1,402,677    <rustc_middle::ty::typeck_results::TypeckResults as rustc_serialize::serialize::Decodable<rustc_middle::query::on_disk_cache::CacheDecoder>>::decode
   -1,317,168    <hashbrown::raw::RawTable<(rustc_span::def_id::LocalDefId, rustc_span::def_id::CrateNum)>>::reserve_rehash::<hashbrown::map::make_hasher<rustc_span::def_id::LocalDefId, rustc_span::def_id::CrateNum, core::hash::BuildHasherDefault<rustc_hash::FxHasher>>::{closure#0}>
    1,317,165    <hashbrown::raw::RawTable<(rustc_ast::node_id::NodeId, rustc_ast::node_id::NodeId)>>::reserve_rehash::<hashbrown::map::make_hasher<rustc_ast::node_id::NodeId, rustc_ast::node_id::NodeId, core::hash::BuildHasherDefault<rustc_hash::FxHasher>>::{closure#0}>
   -1,247,819    <hashbrown::raw::RawTable<(rustc_span::def_id::LocalDefId, alloc::vec::Vec<rustc_span::def_id::DefId>)>>::reserve_rehash::<hashbrown::map::make_hasher<rustc_span::def_id::LocalDefId, alloc::vec::Vec<rustc_span::def_id::DefId>, core::hash::BuildHasherDefault<rustc_hash::FxHasher>>::{closure#0}>
    1,247,819    <hashbrown::raw::RawTable<(rustc_ast::node_id::NodeId, rustc_hir::def::PartialRes)>>::reserve_rehash::<hashbrown::map::make_hasher<rustc_ast::node_id::NodeId, rustc_hir::def::PartialRes, core::hash::BuildHasherDefault<rustc_hash::FxHasher>>::{closure#0}>
     -851,023    rustc_query_system::query::plumbing::ensure_must_run::<rustc_query_impl::DynamicConfig<rustc_query_system::query::caches::VecCache<rustc_span::def_id::LocalDefId, rustc_middle::query::erase::Erased<[u8; 0]>>, false, false, false>, rustc_query_impl::plumbing::QueryCtxt>
      820,268    rustc_query_system::query::plumbing::ensure_must_run::<rustc_query_impl::DynamicConfig<rustc_query_system::query::caches::VecCache<rustc_hir::hir_id::OwnerId, rustc_middle::query::erase::Erased<[u8; 1]>>, false, false, false>, rustc_query_impl::plumbing::QueryCtxt>
      460,984    <rustc_middle::ty::context::CtxtInterners>::intern_ty
     -413,385    <hashbrown::raw::RawTable<(rustc_span::def_id::LocalDefId, ())>>::reserve_rehash::<hashbrown::map::make_hasher<rustc_span::def_id::LocalDefId, (), core::hash::BuildHasherDefault<rustc_hash::FxHasher>>::{closure#0}>
      395,094    <hashbrown::raw::RawTable<(rustc_ast::node_id::NodeId, ())>>::reserve_rehash::<hashbrown::map::make_hasher<rustc_ast::node_id::NodeId, (), core::hash::BuildHasherDefault<rustc_hash::FxHasher>>::{closure#0}>
      352,761    <rustc_expand::mbe::macro_rules::MacroRulesMacroExpander as rustc_expand::base::TTMacroExpander>::expand
     -326,412    <hashbrown::raw::RawTable<(rustc_span::def_id::LocalDefId, rustc_resolve::late::LifetimeUseSet)>>::reserve_rehash::<hashbrown::map::make_hasher<rustc_span::def_id::LocalDefId, rustc_resolve::late::LifetimeUseSet, core::hash::BuildHasherDefault<rustc_hash::FxHasher>>::{closure#0}>
      326,327    <hashbrown::raw::RawTable<(rustc_ast::node_id::NodeId, rustc_resolve::Module)>>::reserve_rehash::<hashbrown::map::make_hasher<rustc_ast::node_id::NodeId, rustc_resolve::Module, core::hash::BuildHasherDefault<rustc_hash::FxHasher>>::{closure#0}>
      161,746    <rustc_ast_lowering::item::ItemLowerer>::lower_node
      130,217    <rustc_resolve::Resolver>::resolve_path_with_ribs
      123,216    <&rustc_middle::ty::list::List<rustc_type_ir::canonical::CanonicalVarInfo<rustc_middle::ty::context::TyCtxt>> as rustc_data_structures::stable_hasher::HashStable<rustc_query_system::ich::hcx::StableHashingContext>>::hash_stable
     -109,488    <rustc_parse::parser::Parser>::parse_expr_assoc_with
     -102,110    <rustc_ast_lowering::LoweringContext>::lower_qpath
      -88,115    rustc_query_system::query::plumbing::try_execute_query::<rustc_query_impl::DynamicConfig<rustc_query_system::query::caches::DefaultCache<rustc_span::def_id::DefId, rustc_middle::query::erase::Erased<[u8; 8]>>, false, false, false>, rustc_query_impl::plumbing::QueryCtxt, true>
       84,242    <rustc_parse::parser::Parser>::parse_path_inner
      -82,416    <hashbrown::raw::RawTable<(rustc_span::def_id::LocalDefId, alloc::vec::Vec<(rustc_middle::hir::place::Place, rustc_middle::mir::syntax::FakeReadCause, rustc_hir::hir_id::HirId)>)>>::reserve_rehash::<hashbrown::map::make_hasher<rustc_span::def_id::LocalDefId, alloc::vec::Vec<(rustc_middle::hir::place::Place, rustc_middle::mir::syntax::FakeReadCause, rustc_hir::hir_id::HirId)>, core::hash::BuildHasherDefault<rustc_hash::FxHasher>>::{closure#0}>
       82,416    <hashbrown::raw::RawTable<(rustc_hir::hir_id::ItemLocalId, alloc::vec::Vec<rustc_middle::ty::Ty>)>>::reserve_rehash::<hashbrown::map::make_hasher<rustc_hir::hir_id::ItemLocalId, alloc::vec::Vec<rustc_middle::ty::Ty>, core::hash::BuildHasherDefault<rustc_hash::FxHasher>>::{closure#0}>
       79,410    rustc_query_system::query::plumbing::try_execute_query::<rustc_query_impl::DynamicConfig<rustc_query_system::query::caches::VecCache<rustc_span::def_id::LocalDefId, rustc_middle::query::erase::Erased<[u8; 8]>>, false, false, false>, rustc_query_impl::plumbing::QueryCtxt, true>
      -63,035    rustc_middle::ty::codec::encode_with_shorthand::<rustc_metadata::rmeta::encoder::EncodeContext, rustc_middle::ty::Ty, <rustc_metadata::rmeta::encoder::EncodeContext as rustc_type_ir::codec::TyEncoder>::type_shorthands>
       60,309    <alloc::sync::Arc<[rustc_span::symbol::Symbol]>>::drop_slow
      -57,726    rustc_data_structures::unord::hash_iter_order_independent::<rustc_query_system::ich::hcx::StableHashingContext, &rustc_span::def_id::LocalDefId, std::collections::hash::set::Iter<rustc_span::def_id::LocalDefId>>
      -52,483    <rustc_infer::infer::InferCtxt>::probe::<core::result::Result<rustc_middle::traits::select::EvaluationResult, rustc_middle::traits::select::OverflowError>, <rustc_trait_selection::traits::select::SelectionContext>::evaluation_probe<<rustc_trait_selection::traits::select::SelectionContext>::evaluate_root_obligation::{closure#0}>::{closure#0}>
       52,188    <rustc_trait_selection::traits::select::SelectionContext>::evaluation_probe::<<rustc_trait_selection::traits::select::SelectionContext>::evaluate_root_obligation::{closure#0}>
       49,326    <rustc_ast_lowering::LoweringContext>::lower_ty_direct
      -47,664    <rustc_resolve::Resolver>::early_resolve_ident_in_lexical_scope
      -38,539    <rustc_parse::parser::Parser>::parse_expr_dot_or_call
       34,695    rustc_query_system::query::plumbing::try_execute_query::<rustc_query_impl::DynamicConfig<rustc_query_system::query::caches::DefaultCache<rustc_span::def_id::DefId, rustc_middle::query::erase::Erased<[u8; 16]>>, false, false, false>, rustc_query_impl::plumbing::QueryCtxt, true>
       33,320    rustc_query_system::query::plumbing::try_execute_query::<rustc_query_impl::DynamicConfig<rustc_query_system::query::caches::DefaultCache<rustc_span::def_id::DefId, rustc_middle::query::erase::Erased<[u8; 24]>>, false, false, false>, rustc_query_impl::plumbing::QueryCtxt, true>
      -31,609    <rustc_resolve::Resolver as rustc_expand::base::ResolverExpand>::resolve_macro_invocation
      -27,378    <rustc_resolve::late::LateResolutionVisitor>::with_lifetime_rib::<(), <rustc_resolve::late::LateResolutionVisitor>::resolve_const_body::{closure#0}>
       27,378    <rustc_resolve::late::LateResolutionVisitor>::resolve_const_body
       25,110    <rustc_lint::levels::TopDown as rustc_lint::levels::LintLevelsProvider>::get_lint_level
       22,697    <rustc_resolve::late::LateResolutionVisitor as rustc_ast::visit::Visitor>::visit_item
```
+ self profile queries show slightly more time spent in `typeck`. Seems like more type checking is happening?

In any case, I don't think that the regression is especially bad, cycles and wall-time are a wash, and this PR does introduce new functionality/feature after all.

**Pull request merged**